### PR TITLE
20.07 - Arch improvements

### PR DESCRIPTION
**Pull request by Scitator**
## Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contribution guide](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md)?
- [ ] Did you check the code style? `catalyst-make-codestyle && catalyst-check-codestyle` (`pip install -U catalyst-codestyle`).
- [ ] Did you make sure to update the docs? We use Google format for all the methods and classes.
- [ ] Did you check the docs with `make check-docs`?
- [ ] Did you write any new necessary tests?
- [ ] Did you add your new functionality to the docs?
- [ ] Did you update the [CHANGELOG](https://github.com/catalyst-team/catalyst/CHANGELOG.md)?
- [x] **You can use 'Login as guest' to see Teamcity build logs.**

## Type of Change

- [ ] Examples / docs / tutorials / contributors update
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves an existing feature)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

File: catalyst/registry/registry.py [hunk 1]
```diff
@@ -142,12 +142,34 @@ def _schedulers_loader(r: Registry):
 SCHEDULERS.late_add(_schedulers_loader)
 Scheduler = SCHEDULERS.add
 
-
 EXPERIMENTS = Registry("experiment")
 Experiment = EXPERIMENTS.add
 
+RUNNERS = Registry("runner")
+Runner = RUNNERS.add
```
Comment (bagxi) [inline catalyst/registry/registry.py]: I'd appreciate it if you would sort registers by importance (e.g. `EXPERIMENTS`, `Experiment`, ..., `RUNNERS`, `Runner`, ..., `_callbacks_loader`, `CALLBACKS`, `Callback`, ...) or alphabetically

File: catalyst/utils/metrics/accuracy.py [hunk 2]
```diff
@@ -13,27 +13,62 @@
 from catalyst.utils.torch import get_activation_fn
 
 
+def multi_class_accuracy(
+    outputs: torch.Tensor, targets: torch.Tensor, topk: Tuple = (1,),
```
Comment (bagxi) [inline catalyst/utils/metrics/accuracy.py]:
```suggestion
    outputs: torch.Tensor, targets: torch.Tensor, topk: Sequence[int] = (1,),
```

File: catalyst/utils/metrics/accuracy.py [hunk 3]
```diff
@@ -13,27 +13,62 @@
...
+) -> List:
```
Comment (bagxi) [inline catalyst/utils/metrics/accuracy.py]:
```suggestion
) -> List[torch.Tensor]:
```

File: catalyst/utils/metrics/accuracy.py [hunk 4]
```diff
@@ -13,27 +13,62 @@
...
+    return output
+
+
+def multi_label_accuracy(
+    outputs: torch.Tensor,
+    targets: torch.Tensor,
+    threshold: float = torch.Tensor,
+) -> torch.Tensor:
+    outputs = (outputs > threshold).long()
+    output = (targets.long() == outputs.long()).sum().float() / np.prod(
```
Comment (bagxi) [inline catalyst/utils/metrics/accuracy.py]: We already converted outputs to `long` type in the line above.
```suggestion
    output = (targets.long() == outputs).sum().float() / np.prod(
```

File: catalyst/utils/metrics/accuracy.py [hunk 5]
```diff
@@ -13,27 +13,62 @@
...
+        targets.shape
+    )
+    return output
+
+
 def accuracy(
     outputs: torch.Tensor,
     targets: torch.Tensor,
+    activation: str = None,
```
Comment (bagxi) [inline catalyst/utils/metrics/accuracy.py]:
```suggestion
    activation: Optional[str] = None,
```

File: catalyst/utils/metrics/accuracy.py [hunk 6]
```diff
@@ -13,27 +13,62 @@
...
+    multi_label: bool = False,
     topk: Tuple = (1,),
     threshold: float = None,
```
Comment (bagxi) [inline catalyst/utils/metrics/accuracy.py]:
```suggestion
    threshold: Optional[float] = None,
```

File: catalyst/utils/metrics/precision.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
+from typing import Tuple
+
+import numpy as np
+
+import torch
+
+
+def average_precision(outputs, targets, k: int = 10):
```
Comment (bagxi) [inline catalyst/utils/metrics/precision.py]: Is it possible to say something about types of the args?

File: catalyst/utils/metrics/precision.py [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+            score += num_hits / (i + 1.0)
+
+    if not targets:
+        return 0.0
+
+    return score / min(len(targets), k)
+
+
+def mean_average_precision(
+    outputs: torch.Tensor, targets: torch.Tensor, topk: Tuple = (1,),
```
Comment (bagxi) [inline catalyst/utils/metrics/precision.py]:
```suggestion
    outputs: torch.Tensor, targets: torch.Tensor, topk: Sequence[int] = (1,),
```

File: catalyst/utils/metrics/precision.py [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+):
+    """Computes the mean average precision at k.
+
+    This function computes the mean average precision at k between two lists
+    of lists of items.
+
+    Args:
+        outputs (list): A list of lists of predicted elements
+        targets (list): A list of lists of elements that are to be predicted
+        topk (int, optional): The maximum number of predicted elements
```
Comment (bagxi) [inline catalyst/utils/metrics/precision.py]: types in docs doesn't match with the annotations

File: catalyst/utils/metrics/precision.py [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+
+    Returns:
+        float: The mean average precision at k over the input lists
+    """
+    max_k = max(topk)
+    _, pred = outputs.topk(max_k, 1, True, True)  # noqa: WPS425
+
+    targets = targets.cpu().detach().numpy().tolist()
```
Comment (bagxi) [inline catalyst/utils/metrics/precision.py]:
```suggestion
    targets = targets.detach().cpu().numpy().tolist()
```

File: catalyst/utils/metrics/precision.py [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+    actual_list = []
+    for a in targets:
+        actual_list.append([a])
+    targets = actual_list
```
Comment (bagxi) [inline catalyst/utils/metrics/precision.py]:
```suggestion
    targets = [[a] for a in targets]
```

**Review by bagxi — DISMISSED** [review]
In my opinion, if we are adding a sub-module for the registry, then it would be a good idea to move `catalyst/catalyst/tools/registry.py` in this sub-module too. If we keep all things related to the registry in one place, it should be a little bit easier to navigate in the project.

File: setup.cfg [hunk 12]
```diff
@@ -1,6 +1,6 @@
 [flake8]
 exclude = .git,__pycache__,docs/conf.py,build,dist
-ignore = C812,C813,C814,C815,C816,D100,D104,D200,D204,D205,D301,D400,D401,D402,D412,D413,DAR003,DAR103,DAR203,E203,E731,E800,E1101,N812,P101,RST201,RST203,RST210,RST213,RST301,RST304,S,W0221,W503,W504,W605,WPS0,WPS100,WPS101,WPS110,WPS111,WPS112,WPS125,WPS2,WPS300,WPS301,WPS305,WPS306,WPS309,WPS317,WPS323,WPS326,WPS331,WPS333,WPS335,WPS336,WPS337,WPS338,WPS342,WPS347,WPS348,WPS349,WPS350,WPS352,WPS402,WPS404,WPS405,WPS408,WPS410,WPS412,WPS414,WPS420,WPS421,WPS425,WPS426,WPS429,WPS430,WPS431,WPS432,WPS433,WPS434,WPS435,WPS440,WPS441,WPS5,WPS6
+ignore = C812,C813,C814,C815,C816,D100,D104,D200,D204,D205,D301,D400,D401,D402,D412,D413,DAR003,DAR103,DAR203,E203,E731,E800,E1101,N812,P101,RST201,RST203,RST210,RST213,RST301,RST304,S,W0221,W503,W504,W605,WPS0,WPS100,WPS101,WPS110,WPS111,WPS112,WPS125,WPS2,WPS300,WPS301,WPS305,WPS306,WPS309,WPS317,WPS323,WPS326,WPS331,WPS333,WPS335,WPS336,WPS337,WPS338,WPS342,WPS347,WPS348,WPS349,WPS350,WPS352,WPS402,WPS404,WPS405,WPS408,WPS410,WPS412,WPS414,WPS420,WPS421,WPS425,WPS426,WPS429,WPS430,WPS431,WPS432,WPS433,WPS434,WPS435,WPS440,WPS441,WPS5,WPS6,WPS345
```
Comment (bagxi) [inline setup.cfg]: In my opinion, it is better to add some `# noqa: WPS345` comments rather than adjusting code style settings (exclude `WPS345`) to fix tests-related issues

File: catalyst/registry/__init__.py [hunk 13]
```diff
@@ -0,0 +1,243 @@
+"""
```
Comment (bagxi) [inline catalyst/registry/__init__.py]: Is it possible to rename `catalyst/registry/registry.py` file to something like `catalyst/registry/registry_factory.py`
and move all logic related to `TRANSFORM`, `Callback`, etc. registries from `.__init__.py` to `catalyst/registry/registry.py`?
